src/main/java/org/jboss/modules/ModuleLoader.java

```diff
@@ -556,17 +575,53 @@ protected final Module findLoadedModuleLocal(String name) {
      * @param module the module to unload
      * @throws SecurityException if an attempt is made to unload a module which does not belong to this module loader
      * @throws SecurityException if the module was not defined by this module loader
+     * @deprecated use {@link #unloadModuleLocal(String,Module)} instead
      */
+    @Deprecated
     protected final void unloadModuleLocal(Module module) throws SecurityException {
+        unloadModuleLocal(null, module);
+    }
+
+    /**
+     * Unload a module from this module loader.  Note that this has no effect on existing modules which refer to the
+     * module being unloaded.  Also, only modules from the current module loader can be unloaded.  Unloading the same
+     * module more than once has no additional effect.  This method only removes the mapping for the module; any running
+     * threads which are currently accessing or linked to the module will continue to function, however attempts to load
+     * this module will fail until a new module is loaded with the same name.  Once this happens, if all references to
+     * the previous module are not cleared, the same module may be loaded more than once, causing possible class duplication
+     * and class cast exceptions if proper care is not taken.
+     *
+     * @param moduleName the module identifier
+     * @param module the module to unload
+     * @return true iff module was unloaded
+     * @throws SecurityException if an attempt is made to unload a module which does not belong to this module loader
+     * @throws SecurityException if the module was not defined by this module loader
+     */
+    protected final boolean unloadModuleLocal(final String moduleName, final Module module) throws SecurityException {
         final ModuleLoader moduleLoader = module.getModuleLoader();
         if (moduleLoader != this) {
             throw new SecurityException("Attempted to unload " + module + " from a different module loader");
         }
-        final String id = module.getIdentifier().toString();
+        final String id = moduleName == null ? module.getIdentifier().toString() : moduleName;
         final FutureModule futureModule = moduleMap.get(id);
         if (futureModule != null && futureModule.module == module) {
-            moduleMap.remove(id, futureModule);
+            final Set<String> aliases = module.aliases;
+            if (aliases != null) {
+                synchronized (aliases) {
+                    if (aliases.remove(id)) {
+                        return true;
+                    } else {
+                        // cleaning aliased module and all its aliases
+                        module.aliases = null;
```

Comment:
If we return `true` here, we have not performed `moduleMap.remove(id, futureModule)` which we should.